src/sentry/utils/concurrent.py

```diff
@@ -4,11 +4,80 @@
 import threading
 from Queue import Full, PriorityQueue
 from concurrent.futures import Future
+from concurrent.futures._base import RUNNING, FINISHED
+from time import time
 
 
 logger = logging.getLogger(__name__)
 
 
+class TimedFuture(Future):
+    def __init__(self, *args, **kwargs):
+        self.__timing = [None, None]  # [started, finished/cancelled]
+        super(TimedFuture, self).__init__(*args, **kwargs)
+
+    def get_timing(self):
+        """\
+        Return the timing data for this future in the form ``(started, finished)``.
+
+        The ``started`` value can be either a timestamp or ``None`` (if the
+        future has not been started.) The ``finished`` value can also be either
+        a timestamp or ``None`` (if the future has not been completed or
+        cancelled.)
+
+        There are some idiosyncracies with the way the timings are recorded:
+
+        - The ``started`` value will generally not be ``None`` if the
+          ``finished`` value is also not ``None``. However, for a future that
+          is marked as cancelled and has yet to be attempted to be executed,
+          the ``finished`` value may be set while the ``started`` value is
+          ``None``.
```

Comment:
Maybe this will become clearer as I read, but above you say

> The ``finished`` value can also be either a timestamp or ``None`` (if the future has not been completed or cancelled.)

So when is finished *not* `None` when it is canceled?

Reply:
Yeah, this reads kind of weird after a second look — I updated it in 6a6d0ac6d62b832761aaa44d49f7ccfe8dd593c9  to hopefully clarify that both completion and cancellation are finished states rather than the ambiguity here.

Reply:
Actually, on third look this description is just not very good. 😅